### PR TITLE
Adding remote watchers to monitor status changes on remote peering

### PR DESCRIPTION
**Pull request by alacuku**
# Description

Added the watchers to monitor the status of resources replicated on remote peering clusters. A remote watcher is triggered when an update is done on a replicated resource living in a remote cluster and reflects those changes on the local cluster where it is running.

Added various labels used to:

- detect which are the resources to be replicated on remote peering clusters
- designate the remote peering cluster where to replicate a given resource
- detect which resources are local ones and which have been replicated from remote peering clusters

Resources to be replicated should have the following labels set:

- `liqo.io/replication: true`
- `liqo.io/remoteID: clusterIDOfThePeeringCluster`

A replicated resource in a remote cluster presents the following labels:

- `liqo.io/replicated: true`
- `liqo.io/originID: clusterIDOfThePeeringClusterThatSentTheResource`

Added the dispatcher operator to the helm chart.

# How Has This Been Tested?

Unit tests for the new functions.
CI tests checking that:

1. creation of a new resource labelled correctly is replicated on the destination peering cluster
2. update on an existing resource is correctly replicated on the destination peering cluster
3. update on the status of a remote resource living in a peering cluster is correctly replicated on the local cluster
4. resources without the correct label are no replicated
5. deletion of an existing resource is correctly performed also on the destination peering cluster

### CODE REVIEW

File: deployments/liqo_chart/subcharts/dispatcher_chart/templates/dispatcher.yaml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+spec:
+  selector:
+    matchLabels:
+      run: dispatcher-operator
+  template:
+    metadata:
+      labels:
+        run: dispatcher-operator
+    spec:
+      tolerations:
```
Comment (palexster) [inline deployments/liqo_chart/subcharts/dispatcher_chart/templates/dispatcher.yaml:97]: Why those tolerations?

Comment (alacuku) [reply deployments/liqo_chart/subcharts/dispatcher_chart/templates/dispatcher.yaml:97]: we don't need them, thanks.

File: internal/dispatcher/dispatcher-operator.go [hunk 2]
```diff
@@ -75,85 +166,207 @@ func (d *DispatcherReconciler) Watcher(gvr schema.GroupVersionResource, stop cha
...
+	namespace := obj.GetNamespace()
+	client := d.LocalDynClient
+	clusterID := d.ClusterID
+	//we check if the resource exists in the local cluster
+	localObj, found, err := d.GetResource(client, gvr, name, namespace, clusterID)
+	if err != nil {
+		klog.Errorf("%s -> an error occurred while getting resource %s of type %s: %s", clusterID, name, gvr.String(), err)
+		return
+	}
+	// TODO if the resource does not exist what do we do?
```
Comment (palexster) [inline internal/dispatcher/dispatcher-operator.go:224]: I would say that since the replicator is the owner of the resource, it can remove the resource on the remote side.

File: internal/dispatcher/dispatcher-operator.go [hunk 3]
```diff
@@ -75,85 +166,207 @@ func (d *DispatcherReconciler) Watcher(gvr schema.GroupVersionResource, stop cha
...
+	//do nothing? remove the remote replication?
+	//current choice -> do nothing
+	if !found {
+		klog.Infof("%s -> resource %s in namespace %s of type %s not found", clusterID, name, namespace, gvr.String())
+		return
+	}
+	//if the resource exists on the local cluster then we update the status
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		//TODO: we reflect on the local resource only the changes of the status of the remote one
```
Comment (palexster) [inline internal/dispatcher/dispatcher-operator.go:233]: This comment is a bit hard to follow

File: internal/dispatcher/dispatcher-operator.go (hunk 2 above)
Comment (alacuku) [reply internal/dispatcher/dispatcher-operator.go:224]: ok.

File: .github/workflows/go.yml [hunk 4]
```diff
@@ -24,6 +24,7 @@ jobs:
         - init-pod-mutator
         - pod-mutator
         - peering-request-webhook-init
+        - dispatcher
```
Comment (frisso) [inline .github/workflows/go.yml:27]: Am I wrong or we came out with a different name for this component?

Comment (alacuku) [reply .github/workflows/go.yml:27]: How should we name it?

Comment (frisso) [reply .github/workflows/go.yml:27]: @palexster Did you take notes? ServiceReplicator or something like that?

Comment (alacuku) [reply .github/workflows/go.yml:27]: Last time we came out with "CRDReplicator" but I am not sure if it will manage only CRDs or also other type of resources. So I kept the old name.

Comment (frisso) [reply .github/workflows/go.yml:27]: That's the reason I used "service" here. Let's wait for Alex.

Comment (palexster) [reply .github/workflows/go.yml:27]: IMHO Service has a pretty clear meaning in Kubernetes and can be misleading referring to service in this context. AFAIK In the near future, this component will handle CRDs replication across clusters, so CRDReplicator looks fine to me.

File: deployments/liqo_chart/subcharts/crdReplicator_chart/Chart.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 0.1.0
+
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+appVersion: 1.16.0
```
Comment (palexster) [inline deployments/liqo_chart/subcharts/crdReplicator_chart/Chart.yaml:23]: 0.1?

**Review by palexster — APPROVED** [review]

File: cmd/crdReplicator/main.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	"k8s.io/client-go/kubernetes"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/klog/v2"
+	"os"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var (
+	scheme          = runtime.NewScheme()
+	clusteIDConfMap = "cluster-id"
```
Comment (fraborg) [inline cmd/crdReplicator/main.go:20]:
```suggestion
	clusterIDConfMap = "cluster-id"
```
typo

File: deployments/liqo_chart/subcharts/crdReplicator_chart/Chart.yaml (hunk 5 above)
Comment (alacuku) [reply deployments/liqo_chart/subcharts/crdReplicator_chart/Chart.yaml:23]: that value is the default one generated by Helm. we should change it for all the subcharts and the umbrella chart as well.

File: cmd/crdReplicator/main.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	if err != nil {
+		klog.Error(err, "unable to start manager")
+		os.Exit(-1)
+	}
+	//create a clientSet
+	k8sClient := kubernetes.NewForConfigOrDie(cfg)
+	//get namespace where the operator is running
+	namespaceName, found := os.LookupEnv("NAMESPACE")
+	if !found {
+		klog.Errorf("namespace env variable not set, please set it in manifest file o the operator")
```
Comment (fraborg) [inline cmd/crdReplicator/main.go:46]:
```suggestion
		klog.Errorf("namespace env variable not set, please set it in manifest file of the operator")
```
typo

File: internal/crdReplicator/crdReplicator-operator.go [hunk 8]
```diff
@@ -0,0 +1,704 @@
...
+			} else {
+				klog.Infof("%s -> dynamic client created", remoteClusterID)
+			}
+			d.RemoteDynClients[remoteClusterID] = dynClient
+		}
+	}
+
+	d.StartWatchers()
+	d.StartRemoteWatchers()
+	d.StopWatchers()
```
Comment (fraborg) [inline internal/crdReplicator/crdReplicator-operator.go:116]: can `defer` on these functions be useful? (in case an error occurs we are sure the watchers are closed)

**Review by fraborg — APPROVED** [review]

File: internal/crdReplicator/crdReplicator-operator.go (hunk 8 above)
Comment (alacuku) [reply internal/crdReplicator/crdReplicator-operator.go:116]: the watchers need to be closed only if the related resource has been removed from the configuration.
if we the secret containing the kubeconfig is not ready it means that watchers for that peering cluster never started so we have nothing to close.

File: cmd/crdReplicator/main.go (hunk 7 above)
Comment (alacuku) [reply cmd/crdReplicator/main.go:46]: thanks
